[MAT-8639] Concurrent Test Case Validation

jkotanchik-SB opened the pull request:
## MADiE PR

Jira Ticket: [MAT-8639](https://jira.cms.gov/browse/MAT-8639)
(Optional) Related Tickets:

### Summary

Implement logic for firing concurrent validation requests of STU 6 test cases. Configuration is controlled Spring config (`application.yml`). Updates validation status to `VALIDATING` when task is picked up by a thread, and again to either `VALID` or `INVALID` once validation (on madie-fhir-service) is complete.

#### Default Executor Config:
- Core Pool Size: 10 (starting thread count)
- Queue Capacity: 500 (how many tasks should be enqueued before more threads are added)
- Max Pool Size: 20 (upper limit of thread count. Increase when queue is full)
- Await Termination: 300 (During a graceful shutdown, how long the service will wait for enqueued tasks to complete, in seconds)


### Additional Changes
- Move test case validation logic into own service. Co-locating the test case validation, async, and concurrency logic prevents circular dependencies.
- Remove talkative debug/testing log message.

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

jkotanchik-SB committed:
MAT-8639: impl multi-thread validation happy path.

jkotanchik-SB committed:
MAT-8639: Move test case validation logic into own service. Co-locating the test case validation, async, and concurrency logic prevents circular dependencies.

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>

jkotanchik-SB committed:
MAT-8639: Remove talkative debug/testing log message.

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>

jkotanchik-SB committed:
MAT-8639: Replace ExecutorService with Spring's ThreadPoolTaskExecutor for its extended functionality. Notably, inspection of queue size and convenience configuration methods.

Log concurrent validation tasks and timings for metric collection.

Update validation status when a task begins execution (VALIDATING) and after receiving a validation result from madie-fhir-service (either VALID or INVALID).

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>

jkotanchik-SB committed:
MAT-8639: Replace separate find and update validation status calls with single, atomic, find and update mongoOperation.

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>

RohitKandimalla commented on src/main/java/cms/gov/madie/measure/services/TestCaseValidationService.java:
```diff
@@ -0,0 +1,194 @@
+            .filter((tc -> tc.getId().equals(submittedTestCase.getId())))
+            .findFirst();
+    if (currentTestCase.isEmpty()) {
+      throw new ResourceNotFoundException("Test Case", submittedTestCase.getId());
```
May be a log ? Hopefully never have to see it, but a golden rule for me is to always have a log when we throw a new exception.

jkotanchik-SB replied:
Ah, yea. I was expecting the log intercepter to log the exception, but there really isn't a guarantee it will get that far back up the chain.

RohitKandimalla commented on src/main/java/cms/gov/madie/measure/services/TestCaseValidationService.java:
```diff
@@ -0,0 +1,194 @@
+            .testCaseValidationStatus(TestCaseValidationStatus.PENDING)
+            .hapiOperationOutcome(null)
+            .build();
+    measure.getTestCases().add(updatedTestCase);
```
I am not liking the idea that this public method assumes that the testCase that needs to be modified is already removed from the list. This should work fine if this method is not public and is not expected to be shared.

RohitKandimalla replied:
May be do findAndUpdateValidationStatus here as well ?

RohitKandimalla commented on src/main/java/cms/gov/madie/measure/services/TestCaseValidationService.java:
```diff
@@ -0,0 +1,194 @@
+    }
+  }
+
+  public HapiOperationOutcome validateTestCaseJson(
```
Now that we are refactoring, I was wondering if we can just expose validateTestCaseAsResource and keep validateTestCaseJson as a private method. If it is possible and if it doesn't add tremendous amount of refactoring.

jkotanchik-SB replied:
Good call. Went with package-private for unit testing.

jkotanchik-SB committed:
MAT-8639: Updates from code review:
- Reduce visibility on json validation method, and refactor calling methods and tests as needed.
- Log exception in addition to throw.
- Replace existing find, modify, save pattern with new findAndUpdate operation to reduce the probability of a race case.

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>

jkotanchik-SB committed:
MAT-8639: Save updated STU6 Test Case before submitting for validation.

Signed-off-by: Joseph Kotanchik <joseph.kotanchik@semanticbits.com>
